include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,1093 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_hdf5_h
+#define dealii_hdf5_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_HDF5
+
+#  include <deal.II/lac/full_matrix.h>
+
+#  include <hdf5.h>
+
+#  include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+// It is necessary to turn clang-format off in order to maintain the Doxygen
+// links because the they are longer than 80 characters
```

Comment:
```suggestion
// links because they are longer than 80 characters
```

Reply:
@bangerth, thanks a lot for your feedback! It improved a lot the code ;)

I reduced the number of commits to 79. I re-ordered and squash the least important commits. It is a large library and I think that keeping the history helps to debug and find bugs. I justified in the body of several commits the choices I made, it would be a pity to loose that documentation.

I added a Changelog entry